src/styles/withStyles.js

```diff
@@ -1,302 +1,29 @@
-          sheetsRegistry.remove(sheetManagerTheme.sheet);
-        }
-      }
-    }
-
-    unsubscribeId = null;
-    jss = null;
-    sheetsManager = sheetsManager;
-    disableStylesGeneration = false;
-    stylesCreatorSaved = null;
-    theme = null;
-    sheetOptions = null;
-    theme = null;
-
-    render() {
-      const { classes: classesProp, innerRef, ...other } = this.props;
-
-      let classes;
-      let renderedClasses = {};
-
-      if (!this.disableStylesGeneration) {
-        const sheetManager = this.sheetsManager.get(this.stylesCreatorSaved);
-        const sheetsManagerTheme = sheetManager.get(this.theme);
-        renderedClasses = sheetsManagerTheme.sheet.classes;
-      }
-
-      if (classesProp) {
-        classes = {
-          ...renderedClasses,
-          ...Object.keys(classesProp).reduce((accumulator, key) => {
-            warning(
-              renderedClasses[key] || this.disableStylesGeneration,
-              [
-                `Material-UI: the key \`${key}\` ` +
-                  `provided to the classes property is not implemented in ${getDisplayName(
-                    Component,
-                  )}.`,
-                `You can only override one of the following: ${Object.keys(renderedClasses).join(
-                  ',',
-                )}`,
-              ].join('\n'),
-            );
-
-            warning(
-              !classesProp[key] || typeof classesProp[key] === 'string',
-              [
-                `Material-UI: the key \`${key}\` ` +
-                  `provided to the classes property is not valid for ${getDisplayName(Component)}.`,
-                `You need to provide a non empty string instead of: ${classesProp[key]}.`,
-              ].join('\n'),
-            );
-
-            if (classesProp[key]) {
-              accumulator[key] = `${renderedClasses[key]} ${classesProp[key]}`;
-            }
-
-            return accumulator;
-          }, {}),
-        };
-      } else {
-        classes = renderedClasses;
-      }
-
-      const more = {};
-
-      // Provide the theme to the wrapped component.
-      // So we don't have to use the `withTheme()` Higher-order Component.
-      if (withTheme) {
-        more.theme = this.theme;
-      }
-
-      return <Component classes={classes} {...more} {...other} ref={innerRef} />;
-    }
-  }
+  const Style = props => {
+    const { classes, innerRef, ...componentProps } = props;
+    return (
+      <Styled classes={classes} Component={Component} {...options} styles={stylesOrCreator}>
+        {styledProps => {
+          const { theme, ...rest } = styledProps;
+          const styledPropsWithOptions = rest;
+          if (options.withTheme) {
+            styledPropsWithOptions.theme = theme;
+          }
+          if (innerRef) {
+            styledPropsWithOptions.ref = innerRef;
+          }
+          return <Component {...styledPropsWithOptions} {...componentProps} />;
+        }}
+      </Styled>
+    );
+  };
```

Comment:
This is an example where the render callback pattern is better.  `withTheme` and `innerRef` are HOC options not necessary using `Styled`.  With a render callback pattern, the child chooses what it uses and what it ignores (e.g. `theme`), and the `innerRef` indirection is unnecessary because the parent can put the `ref` where ever it chooses.  This code shows what it took to match the various current usage patterns of the HOC.